### PR TITLE
User stats api and tabs

### PR DESCRIPTION
**Pull request by amrondonp**
User statistics API and frontend code for vizualization

@0xDCA @larranaga @Maikol1804 Ready for review

### CODE REVIEW

File: inginious/frontend/webapp/plugins/statistics/pages.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                .user_statistics(course_id)
+        )
+
+
+class UserStatisticsAPI(INGIniousAuthPage):
+    def GET_AUTH(self, *args, **kwargs):
+        self.ensure_parameters()
+        return self.statistics()
+
+    def ensure_parameters(self):
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/pages.py]: This name sounds weird. What about something like validate_parameters ?

File: inginious/frontend/webapp/plugins/statistics/pages.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        username = self.user_manager.session_username()
+        course_id = web.input(course_id=None).course_id
+
+        if course_id is None:
+            raise web.badrequest("400 Bad Request: Missing course_id in the query params")
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/pages.py]: Send a JSON instead

File: inginious/frontend/webapp/plugins/statistics/pages.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+        try:
+            course = self.course_factory.get_course(course_id)
+        except:
+            raise web.notfound("404 Not found: The course does not exist")
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/pages.py]: Send a JSON instead

File: inginious/frontend/webapp/plugins/statistics/pages.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+        if not self.user_manager.course_is_user_registered(course, username):
+            raise web.forbidden("403 Forbidden: You are not registered in this course")
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/pages.py]: I would recommend merging this with the 404 case. Users should not be able to make requests to this API to check if a course exists or not, so please return 404 when the course does not exist or the user does not have permissions (and change the message to something like "The course does not exist or the user does not have permissions")

File: inginious/frontend/webapp/plugins/statistics/pages.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        self.template_helper.add_javascript("/static/statistics/js/statistics.js")
+
+        return (
+            self.template_helper
+                .get_custom_renderer(_BASE_RENDERER_PATH)
+                .user_statistics(course_id)
+        )
+
+
+class UserStatisticsAPI(INGIniousAuthPage):
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/pages.py]: Consider inheriting from inginious.frontend.webapp.pages.api._api_page.APIPage, which has some helpers to use JSON for errors, and also to handle API requests that do not use the appropiate HTTP method (GET, POST, etc.)

File: inginious/frontend/webapp/plugins/statistics/pages.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            raise web.notfound("404 Not found: The course does not exist")
+
+        if not self.user_manager.course_is_user_registered(course, username):
+            raise web.forbidden("403 Forbidden: You are not registered in this course")
+
+    def statistics(self):
+        return "[]"
+
+
+class TrialsAndBestGrade(UserStatisticsAPI):
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/pages.py]: Add Api at the end of the name to keep the class name clear

File: inginious/frontend/webapp/plugins/statistics/pages.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                    {
+                        "result": {"$ne": None}
+                    }
+            }
+        ])
+
+        return json.dumps(list(best_submissions), cls=DateTimeEncoder)
+
+
+class DateTimeEncoder(json.JSONEncoder):
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/pages.py]: Shall we move this to another file?

File: inginious/frontend/webapp/plugins/statistics/pages.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    def default(self, o):
+        if isinstance(o, datetime):
+            return o.isoformat()
+
+
+class BarSubmissionsPerTasks(UserStatisticsAPI):
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/pages.py]: Add Api at the end of the name to keep the class name clear

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+    };
+
+    Statistic.prototype._fetchData = function () {
+        throw 'Not implemented';
+    };
+
+    return Statistic;
+})();
+
+var ExampleStatistic = (function () {
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Remove this, please

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+                    "text": [],
+                    "x": [],
+                    "y": []
+                }
+        }
+
+
+        for (var index = 0; index < tries_per_tasks.length; index++) {
+            for (var j = 0; j < results.length; j++) {
+                if (tries_per_tasks[index].result === results[j]) {
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: This would be a lot clearer if you used indexOf instead

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+            data.push(plotData[results[index]]);
+        }
+
+        var layout = {
+            xaxis: {title: 'Task'},
+            yaxis: {title: 'Grade', range: [-10, 110]},
+            margin: {t: 20},
+            hovermode: 'closest'
+        };
+        Plotly.purge(this.div_id);
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Why is this needed?

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        for (var i = 0; i < data.length; ++i) {
+            if (data_count_obj[data[i].task_id] != null) {
+                data_count_obj[data[i].task_id] += data[i].count;
+            } else {
+                data_count_obj[data[i].task_id] = data[i].count;
+                tasks_id.push(data[i].task_id)
+            }
+        }
+
+        var compilation_error_data = this.createObjectToPlotData(data, data_count_obj, "COMPILATION_ERROR", 'rgb(236,199,6)');
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Abstract the colors in constants so they can be reused

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        this.normalize = !this.normalize;
+        this.plotAsync();
+    };
+
+    BarSubmissionsPerTasks.prototype._plotData = function (data) {
+        var data_count_obj = {};
+        var tasks_id = [];
+
+        for (var i = 0; i < data.length; ++i) {
+            if (data_count_obj[data[i].task_id] != null) {
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Consider reversing the order of these two branches (positive case first). It's a bit easier to read something like this:

```
if (data_count_obj[data[i].task_id] == null) {
    data_count_obj[data[i].task_id] = data[i].count;
    tasks_id.push(data[i].task_id);
} else {
    data_count_obj[data[i].task_id] += data[i].count;
}
```

Actually, it's more elegant if you merge the two cases like this:

```
if (data_count_obj[data[i].task_id] == null) {
    data_count_obj[data[i].task_id] = 0;
    tasks_id.push(data[i].task_id);
}

data_count_obj[data[i].task_id] += data[i].count;
```

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+                data_count_obj[data[i].task_id] += data[i].count;
+            } else {
+                data_count_obj[data[i].task_id] = data[i].count;
+                tasks_id.push(data[i].task_id)
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Don't forget the semicolon

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+            yaxis: {
+                title: this.normalize ? 'Percentage of submissions' : 'Number of submissions',
+                titlefont: {
+                    size: 16,
+                    color: 'rgb(107, 107, 107)'
+                }
+            }
+        };
+
+        Plotly.purge(this.id_div);
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Why is this needed? (x2)

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+            y: [],
+            marker: {color: color_category},
+            name: verdict,
+            type: 'bar'
+        };
+
+        for (var i = 0; i < data.length; ++i) {
+            if (data[i].summary_result === verdict) {
+                plotData.x.push(data[i].task_id);
+                if (!this.normalize) {
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: It's a bit clearer if you invert the branch order so the positive case is first ( `if (this.normalize) {` )

File: inginious/frontend/webapp/plugins/statistics/user_statistics.html [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    <div id="trials-circle" class="tab-pane fade in active">
+        <h3>Trials and Best Grade</h3>
+        <div id="tries_per_task" style="width: 100%"><!-- Plotly chart will be drawn inside this DIV --></div>
+        <a href='#' onclick='UserStatistics.trialsAndBestGradeStatistic.downloadCsvAsync();'>Download CSV</a>
+    </div>
+    <div id="bar-submissions-per-tasks" class="tab-pane fade">
+        <h3>Bar Submissions Per Tasks</h3>
+        <div id="submissions_per_task" style="height: 500px;"><!-- Plotly chart will be drawn inside this DIV --></div>
+        <div class="normalize" id="normalize_submissions_per_task" style="display: flex;justify-content: center;">
+          <button style="background-color: white;padding: 8px 15px;text-align: center;font-size: 12px;"
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/user_statistics.html]: It's clearer if you move the styles to the CSS, and use CSS classes

File: inginious/frontend/webapp/plugins/statistics/user_statistics.html [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+</div>
+
+
+<script>
+    $$(function () {
+        UserStatistics.trialsAndBestGradeStatistic = new UserTrialsAndBestGradeStatistic("$:course_id");
+        UserStatistics.barSubmissionsPerTasks = new BarSubmissionsPerTasks("$:course_id");
+
+        $$('a[data-toggle="tab"]').on('shown.bs.tab', function (e) {
+            if(e.target.id === "trials-circle-tab"){
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/user_statistics.html]: Consider doing something like this: https://github.com/JuezUN/INGInious/pull/33/files#diff-561c95b01b3fcf06768329ce8e84e4f8R101

**Comment by larranaga** [comment]
lgtm

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js (hunk 10 above)
Comment (amrondonp) [reply inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: I found that the inner for was unnesesary! I don't know why i put it in the first place

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js (hunk 11 above)
Comment (amrondonp) [reply inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: image
Each click tab plots the statistic again, so this would happen if we didn't purge it before

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js (hunk 15 above)
Comment (amrondonp) [reply inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: It's a similar reason for what we did in the previous one

File: inginious/frontend/webapp/plugins/statistics/pages.py (hunk 5 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/pages.py]: Done

File: inginious/frontend/webapp/plugins/statistics/pages.py (hunk 1 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/pages.py]: Done

File: inginious/frontend/webapp/plugins/statistics/pages.py (hunk 2 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/pages.py]: Done

File: inginious/frontend/webapp/plugins/statistics/pages.py (hunk 3 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/pages.py]: Done

File: inginious/frontend/webapp/plugins/statistics/pages.py (hunk 4 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/pages.py]: Done

File: inginious/frontend/webapp/plugins/statistics/pages.py (hunk 6 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/pages.py]: Done

File: inginious/frontend/webapp/plugins/statistics/pages.py (hunk 7 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/pages.py]: Unnecessary now

File: inginious/frontend/webapp/plugins/statistics/pages.py (hunk 8 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/pages.py]: Done

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js (hunk 9 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Done

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js (hunk 11 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Thanks

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js (hunk 13 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Done

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js (hunk 14 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Done

File: inginious/frontend/webapp/plugins/statistics/static/js/statistics.js (hunk 12 above)
Comment (0xDCA) [reply inginious/frontend/webapp/plugins/statistics/static/js/statistics.js]: Done

File: inginious/frontend/webapp/plugins/statistics/static/js/user_statistics.js [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+        this.normalize = false;
+    }
+
+    BarSubmissionsPerTasks.prototype = Object.create(Statistic.prototype);
+
+    BarSubmissionsPerTasks.prototype._fetchData = function () {
+        return $.getJSON(this.RESOURCE_URL, {course_id: this.course_id});
+    };
+
+    BarSubmissionsPerTasks.prototype.toggle_normalize = function () {
```
Comment (0xDCA) [inline inginious/frontend/webapp/plugins/statistics/static/js/user_statistics.js]: nit: use camelCase to be consistent with the rest of methods

**Review by 0xDCA — APPROVED** [review]
